web/modules/custom/custom_page/src/Plugin/Block/TableOfContentsOutline.php

```diff
@@ -0,0 +1,368 @@
+   *
+   * Recursively check the structured menu tree in order to obtain the parent
+   * link.
+   *
+   * @param \Drupal\Core\Menu\MenuLinkTreeElement[] $tree
+   *   The menu tree or subtree to check for the parent.
+   * @param string $active_link_id
+   *   The active link id.
+   * @param \Drupal\Core\Menu\MenuLinkTreeElement $parent
+   *   The current parent found.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement|null
+   *   The menu link tree.
+   */
+  protected function getParentElement(array $tree, string $active_link_id, MenuLinkTreeElement $parent = NULL): ?MenuLinkTreeElement {
+    if (empty($tree)) {
+      return NULL;
+    }
+
+    if (isset($tree[$active_link_id])) {
+      return $parent;
+    }
+
+    foreach ($tree as $menu_link_id => $menu_item) {
+      if ($parent = $this->getParentElement($tree[$menu_link_id]->subtree, $active_link_id, $menu_item)) {
+        return $parent;
+      }
+    }
+
+    // If none of the above have found the parent, then the link was not found.
+    return $parent;
+  }
+
+  /**
+   * Loads and returns the og menu instance.
+   *
+   * @return \Drupal\og_menu\OgMenuInstanceInterface
+   *   The og menu instance.
+   */
+  protected function getOgMenuInstance(): OgMenuInstanceInterface {
+    if (empty($this->ogMenuInstance)) {
+      $this->ogMenuInstance = $this->ogMenuManager->getOgMenuInstanceByCustomPage($this->node);
+    }
+
+    return $this->ogMenuInstance;
+  }
+
+  /**
+   * Returns the active link.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkInterface
+   *   The active link.
+   */
+  protected function getActiveLink(): MenuLinkInterface {
+    if (empty($this->activeLink)) {
+      $og_menu_id = $this->getOgMenuName();
+      $this->activeLink = $this->activeTrail->getActiveLink($og_menu_id);
+    }
+
+    return $this->activeLink;
+  }
+
+  /**
+   * Returns the menu tree of the current og menu.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement[]
+   *   The og menu tree.
+   */
+  protected function getMenuTree(): array {
+    if (empty($this->menuTree)) {
+      $og_menu_id = $this->getOgMenuName();
+      $this->menuTree = $this->menuLinkTree->load($og_menu_id, new MenuTreeParameters());
+    }
+
+    return $this->menuTree;
+  }
+
+  /**
+   * Retrieves the og menu in a flattened list.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement[]
+   *   The flattened menu.
+   */
+  protected function getFlattenedMenu(): array {
+    if (empty($this->flattenedMenu)) {
+      $og_menu_id = $this->getOgMenuName();
+      $tree = $this->menuLinkTree->load($og_menu_id, new MenuTreeParameters());
+      $this->flatOutlineTree($tree);
+
+      foreach ($this->flattenedMenu as $menu_link_id => $menu_item) {
+        $title = $menu_item->getTitle();
+        /** @var \Drupal\Core\Url $url */
+        $route_name = $menu_item->getRouteName();
+
+        if ($title == 'Overview' && $route_name === 'entity.rdf_entity.canonical') {
```

Comment:
Maybe we can skip loading these items at all in the `getMenuTree()` method using the tree parameters?
Or at least avoid to have these 3 strings in 3 different places.